File: ArtFruit/Sources/WallpaperService.swift
import AppKit

@MainActor
final class WallpaperService {
    private let fileManager = FileManager.default
    private let cacheDir: URL
    private var lastWallpaperURLs: [URL] = []

    init() {
        cacheDir = fileManager.temporaryDirectory
            .appendingPathComponent("io.github.bpiche.ArtFruit", isDirectory: true)
        try? fileManager.createDirectory(at: cacheDir, withIntermediateDirectories: true)
        NSLog("[ArtFruit] Cache dir: \(cacheDir.path)")
    }

    /// Apply an artwork to the given screens (defaults to all screens when `screens` is nil).
    func apply(imageURL: URL, title: String, artist: String, showTitle: Bool = true, showArtist: Bool = true, screens: [NSScreen]? = nil) async throws {
        NSLog("[ArtFruit] Downloading image from \(imageURL)...")
        let (data, _) = try await URLSession.shared.data(from: imageURL)
        NSLog("[ArtFruit] Downloaded \(data.count) bytes.")

        guard let sourceImage = NSImage(data: data) else {
            throw WallpaperError.invalidImage
        }
        NSLog("[ArtFruit] Source image size: \(sourceImage.size)")

        let screens = screens ?? NSScreen.screens
        NSLog("[ArtFruit] Applying to \(screens.count) screen(s)...")

        let timestamp = Int(Date().timeIntervalSince1970)
        var newURLs: [URL] = []

        for (i, screen) in screens.enumerated() {
            // Use backing pixel size so retina screens get full-res images
            let screenRect = screen.frame
            let scale = screen.backingScaleFactor
            let pixelWidth = Int(screenRect.width * scale)
            let pixelHeight = Int(screenRect.height * scale)
            NSLog("[ArtFruit] Screen \(i): \(pixelWidth)×\(pixelHeight)px (scale \(scale)x)")

            let fitted = try fitImage(sourceImage, toWidth: pixelWidth, height: pixelHeight,
                                      title: title, artist: artist,
                                      showTitle: showTitle, showArtist: showArtist)
            let url = cacheDir.appendingPathComponent("artfruit_\(timestamp)_screen\(i).jpg")
            try saveJPEG(fitted, to: url)
            newURLs.append(url)

            do {
                // Image is already composited to exact pixel dimensions — no extra scaling needed
                try NSWorkspace.shared.setDesktopImageURL(url, for: screen, options: [:])
                NSLog("[ArtFruit] Screen \(i): applied \(pixelWidth)×\(pixelHeight) wallpaper")
            } catch {
                NSLog("[ArtFruit] Screen \(i): FAILED: \(error.localizedDescription)")
                throw error
            }
        }

        // Clean up previous wallpaper files
        for old in lastWallpaperURLs {
            try? fileManager.removeItem(at: old)
        }
        lastWallpaperURLs = newURLs
        NSLog("[ArtFruit] All screens updated.")
    }

    /// Scale-to-fill (cover): artwork fills the canvas, cropping if needed.
    /// Adds a blurred/darkened background so portrait art on landscape screens looks good.
    private func fitImage(_ source: NSImage, toWidth width: Int, height: Int,
                          title: String, artist: String,
                          showTitle: Bool, showArtist: Bool) throws -> NSImage {
        let canvasSize = NSSize(width: width, height: height)
        let srcSize = source.size
        guard srcSize.width > 0, srcSize.height > 0 else { throw WallpaperError.invalidImage }

        let result = NSImage(size: canvasSize)
        guard let rep = NSBitmapImageRep(
            bitmapDataPlanes: nil,
            pixelsWide: width,
            pixelsHigh: height,
            bitsPerSample: 8,
            samplesPerPixel: 4,
            hasAlpha: true,
            isPlanar: false,
            colorSpaceName: .deviceRGB,
            bytesPerRow: 0,
            bitsPerPixel: 0
        ) else { throw WallpaperError.encodeFailure }
        result.addRepresentation(rep)

        result.lockFocus()
        defer { result.unlockFocus() }

        let canvas = NSRect(origin: .zero, size: canvasSize)

        // --- Background: fill with black, then draw a blurred/dimmed version of the
        //     artwork stretched to cover (so letterbox bars aren't just black)
        NSColor.black.setFill()
        canvas.fill()

        // Draw blurred background (stretched to fill, then dimmed)
        let bgScale = max(CGFloat(width) / srcSize.width, CGFloat(height) / srcSize.height)
        let bgW = srcSize.width * bgScale
        let bgH = srcSize.height * bgScale
        let bgRect = NSRect(
            x: (CGFloat(width) - bgW) / 2,
            y: (CGFloat(height) - bgH) / 2,
            width: bgW,
            height: bgH
        )
        source.draw(in: bgRect, from: .zero, operation: .copy, fraction: 0.35)

        // --- Foreground: scale-to-fit (letterbox) centered
        let fitScale = min(CGFloat(width) / srcSize.width, CGFloat(height) / srcSize.height)
        let fgW = srcSize.width * fitScale
        let fgH = srcSize.height * fitScale
        let fgRect = NSRect(
            x: (CGFloat(width) - fgW) / 2,
            y: (CGFloat(height) - fgH) / 2,
            width: fgW,
            height: fgH
        )
        source.draw(in: fgRect, from: .zero, operation: .sourceOver, fraction: 1.0)

        // --- Title/artist overlay (bottom-right of the artwork rect)
        var textParts: [String] = []
        if showTitle { textParts.append(title) }
        if showArtist { textParts.append(artist) }
        let text = textParts.joined(separator: "  ")

        if !text.isEmpty {
            let fontSize: CGFloat = max(24, min(44, CGFloat(width) / 60))
            let textAttrs: [NSAttributedString.Key: Any] = [
                .font: NSFont.systemFont(ofSize: fontSize, weight: .medium),
                .foregroundColor: NSColor.white
            ]
            let shadowAttrs: [NSAttributedString.Key: Any] = [
                .font: NSFont.systemFont(ofSize: fontSize, weight: .medium),
                .foregroundColor: NSColor.black.withAlphaComponent(0.75)
            ]
            let measured = NSAttributedString(string: text, attributes: textAttrs).size()
            let padding: CGFloat = max(16, CGFloat(width) / 80)
            let tx = fgRect.maxX - measured.width - padding
            let ty = fgRect.minY + padding

            NSAttributedString(string: text, attributes: shadowAttrs)
                .draw(at: NSPoint(x: tx + 1, y: ty - 1))
            NSAttributedString(string: text, attributes: textAttrs)
                .draw(at: NSPoint(x: tx, y: ty))
        }

        return result
    }

    private func saveJPEG(_ image: NSImage, to url: URL) throws {
        guard let tiffData = image.tiffRepresentation,
              let bitmap = NSBitmapImageRep(data: tiffData),
              let jpegData = bitmap.representation(using: .jpeg, properties: [.compressionFactor: 0.92])
        else { throw WallpaperError.encodeFailure }
        try jpegData.write(to: url)
        NSLog("[ArtFruit] Saved \(url.lastPathComponent) (\(jpegData.count) bytes)")
    }

    enum WallpaperError: LocalizedError {
        case invalidImage, encodeFailure
        var errorDescription: String? {
            switch self {
            case .invalidImage: return "Downloaded data is not a valid image."
            case .encodeFailure: return "Could not encode wallpaper image."
            }
        }
    }
}

File: artfruit_windows/WallpaperService.cs
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Drawing.Text;
using System.Net.Http;

namespace ArtFruit;

/// <summary>
/// A single monitor to paint: its pixel bounds plus, when available, the shell
/// device id used to target it via <see cref="DesktopWallpaper"/>.
/// </summary>
public readonly record struct MonitorTarget(Rectangle PixelBounds, string? DeviceId);

/// <summary>
/// Downloads artwork and composites a per-monitor wallpaper (black fill →
/// dimmed/stretched background → scale-to-fit foreground → title/artist overlay),
/// then applies it. Ports the Swift <c>WallpaperService</c>.
/// </summary>
public sealed class WallpaperService
{
    private readonly HttpClient _http;
    private readonly string _cacheDir;
    private List<string> _lastWallpaperFiles = new();

    public WallpaperService(HttpClient http)
    {
        _http = http;
        _cacheDir = Path.Combine(Path.GetTempPath(), "io.github.bpiche.ArtFruit");
        Directory.CreateDirectory(_cacheDir);
        Log.Info($"Cache dir: {_cacheDir}");
    }

    /// <summary>
    /// Applies <paramref name="artwork"/> to the given targets (defaults to all
    /// monitors when <paramref name="targets"/> is null). The same artwork is
    /// fitted independently to each target's pixel resolution.
    /// </summary>
    public async Task ApplyAsync(
        Artwork artwork,
        bool showTitle,
        bool showArtist,
        IReadOnlyList<MonitorTarget>? targets = null,
        CancellationToken ct = default)
    {
        Log.Info($"Downloading image from {artwork.ImageUrl}...");
        var data = await _http.GetByteArrayAsync(artwork.ImageUrl, ct).ConfigureAwait(false);
        Log.Info($"Downloaded {data.Length} bytes.");

        using var ms = new MemoryStream(data);
        using var sourceImage = new Bitmap(ms);
        Log.Info($"Source image size: {sourceImage.Width}x{sourceImage.Height}");

        var monitorTargets = targets ?? Monitors.Enumerate();
        Log.Info($"Applying to {monitorTargets.Count} screen(s)...");

        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        var newFiles = new List<string>();

        for (var i = 0; i < monitorTargets.Count; i++)
        {
            var target = monitorTargets[i];
            var pixelWidth = Math.Max(target.PixelBounds.Width, 1);
            var pixelHeight = Math.Max(target.PixelBounds.Height, 1);
            Log.Info($"Screen {i}: {pixelWidth}x{pixelHeight}px");

            using var fitted = FitImage(sourceImage, pixelWidth, pixelHeight,
                artwork.Title, artwork.Artist, showTitle, showArtist);

            var file = Path.Combine(_cacheDir, $"artfruit_{timestamp}_screen{i}.jpg");
            SaveJpeg(fitted, file);
            newFiles.Add(file);

            ApplyToMonitor(target, file, i);
        }

        // Clean up previous wallpaper files.
        foreach (var old in _lastWallpaperFiles)
        {
            try { File.Delete(old); } catch { /* best effort */ }
        }
        _lastWallpaperFiles = newFiles;
        Log.Info("All screens updated.");
    }

    private static void ApplyToMonitor(MonitorTarget target, string file, int index)
    {
        try
        {
            if (target.DeviceId is { Length: > 0 } id)
            {
                DesktopWallpaper.SetWallpaper(id, file);
                Log.Info($"Screen {index}: applied wallpaper via IDesktopWallpaper");
            }
            else
            {
                LegacyWallpaper.SetForAllMonitors(file);
                Log.Info($"Screen {index}: applied wallpaper via SystemParametersInfo (all monitors)");
            }
        }
        catch (Exception ex)
        {
            Log.Info($"Screen {index}: FAILED: {ex.Message}");
            // Last-ditch fallback so the user still gets *a* wallpaper.
            try { LegacyWallpaper.SetForAllMonitors(file); } catch { /* give up */ }
        }
    }

    /// <summary>
    /// Scale-to-fit (letterbox) the artwork centered on a canvas, over a dimmed,
    /// scale-to-fill blurred background, with an optional title/artist overlay.
    /// </summary>
    private static Bitmap FitImage(
        Image source, int width, int height,
        string title, string artist, bool showTitle, bool showArtist)
    {
        var result = new Bitmap(width, height, PixelFormat.Format32bppArgb);
        using var g = Graphics.FromImage(result);
        g.InterpolationMode = InterpolationMode.HighQualityBicubic;
        g.SmoothingMode = SmoothingMode.HighQuality;
        g.PixelOffsetMode = PixelOffsetMode.HighQuality;
        g.TextRenderingHint = TextRenderingHint.AntiAliasGridFit;

        // --- Background: black fill, then a dimmed scale-to-fill copy of the art.
        g.Clear(Color.Black);

        double srcW = source.Width;
        double srcH = source.Height;

        var bgScale = Math.Max(width / srcW, height / srcH);
        var bgW = (float)(srcW * bgScale);
        var bgH = (float)(srcH * bgScale);
        var bgRect = new RectangleF(
            (float)((width - bgW) / 2),
            (float)((height - bgH) / 2),
            bgW, bgH);

        // Dim to 35% opacity (matches the Swift fraction: 0.35).
        using (var dim = new ImageAttributes())
        {
            var matrix = new ColorMatrix { Matrix33 = 0.35f };
            dim.SetColorMatrix(matrix, ColorMatrixFlag.Default, ColorAdjustType.Bitmap);
            g.DrawImage(
                source,
                new[]
                {
                    new PointF(bgRect.Left, bgRect.Top),
                    new PointF(bgRect.Right, bgRect.Top),
                    new PointF(bgRect.Left, bgRect.Bottom),
                },
                new RectangleF(0, 0, source.Width, source.Height),
                GraphicsUnit.Pixel,
                dim);
        }

        // --- Foreground: scale-to-fit (letterbox), centered.
        var fitScale = Math.Min(width / srcW, height / srcH);
        var fgW = (float)(srcW * fitScale);
        var fgH = (float)(srcH * fitScale);
        var fgRect = new RectangleF(
            (float)((width - fgW) / 2),
            (float)((height - fgH) / 2),
            fgW, fgH);
        g.DrawImage(source, fgRect);

        // --- Title/artist overlay (bottom-right of the artwork rect).
        var parts = new List<string>();
        if (showTitle) parts.Add(title);
        if (showArtist) parts.Add(artist);
        var text = string.Join("  ", parts);

        if (!string.IsNullOrWhiteSpace(text))
        {
            var fontSize = (float)Math.Max(24, Math.Min(44, width / 60.0));
            using var font = new Font("Segoe UI", fontSize, FontStyle.Regular, GraphicsUnit.Pixel);
            var measured = g.MeasureString(text, font);
            var padding = (float)Math.Max(16, width / 80.0);

            // GDI+ origin is top-left, so bottom padding is measured from fgRect.Bottom.
            var tx = fgRect.Right - measured.Width - padding;
            var ty = fgRect.Bottom - measured.Height - padding;

            using var shadow = new SolidBrush(Color.FromArgb(191, 0, 0, 0)); // ~0.75 alpha
            using var white = new SolidBrush(Color.White);
            g.DrawString(text, font, shadow, tx + 1, ty + 1);
            g.DrawString(text, font, white, tx, ty);
        }

        return result;
    }

    private static void SaveJpeg(Bitmap image, string path)
    {
        var encoder = ImageCodecInfo.GetImageEncoders()
            .First(c => c.FormatID == ImageFormat.Jpeg.Guid);
        using var parameters = new EncoderParameters(1);
        parameters.Param[0] = new EncoderParameter(Encoder.Quality, 92L);
        image.Save(path, encoder, parameters);
        Log.Info($"Saved {Path.GetFileName(path)}");
    }
}
